### PR TITLE
Removing DefaultHostingConfiguration

### PR DESCRIPTION
**Pull request by JunTaoLuo**
This issue addresses #727, #700 and render #693 not applicable

As discussed offline with @davidfowl @Tratcher , we are removing default hosting configurations and the user is expected to construct their own configurations. The only exception is that we fallback to environment variables for the `Environment` option. I'll make an update to IISIntegration in a later PR to see what the changes there look like.

We will also need to update all of our samples again.

### CODE REVIEW

File: samples/SampleStartups/StartupHelloWorld.cs [hunk 1]
```diff
@@ -27,8 +28,10 @@ public void Configure(IApplicationBuilder app)
         // Entry point for the application.
         public static void Main(string[] args)
         {
+            var config = new ConfigurationBuilder().AddCommandLine(args).Build();
```
Comment (davidfowl) [inline samples/SampleStartups/StartupHelloWorld.cs]: Remove config from this sample.

File: test/Microsoft.AspNetCore.Hosting.Tests/WebHostBuilderTests.cs [hunk 2]
```diff
@@ -353,6 +352,41 @@ public void UseEnvironmentIsNotOverriden()
...
+            // Default to Production
+            var host = new WebHostBuilder()
+                .UseServer(testServer)
+                .UseStartup("Microsoft.AspNetCore.Hosting.Tests")
+                .Build();
+
+            Assert.Equal(EnvironmentName.Production, host.Services.GetService<IHostingEnvironment>().EnvironmentName);
+
+            // Use environment variables if available
+            Environment.SetEnvironmentVariable(WebHostDefaults.EnvironmentVariablesPrefix + WebHostDefaults.EnvironmentKey, "EnvironmentA");
```
Comment (davidfowl) [inline test/Microsoft.AspNetCore.Hosting.Tests/WebHostBuilderTests.cs]: Scary, let's *not* have a test for this. Global state like this affects tests running in parallel.

Comment (JunTaoLuo) [inline test/Microsoft.AspNetCore.Hosting.Tests/WebHostBuilderTests.cs]: Maybe we can launch the test in a subprocess with environment variables set so it doesn't affect the global state of the current process? I can also just remove it if that's too much trouble.

Comment (Tratcher) [inline test/Microsoft.AspNetCore.Hosting.Tests/WebHostBuilderTests.cs]: The test deployers already set the environment this way, correct? That should be adequate coverage.

**Comment by Tratcher** [comment]
Side note: The IISIntegration changes need to be checked in first, and they don't rely on any of these changes.

File: samples/SampleStartups/StartupFullControl.cs [hunk 3]
```diff
@@ -13,7 +14,14 @@ public class StartupFullControl
     {
         public static void Main(string[] args)
         {
+            var config = new ConfigurationBuilder()
+                .AddCommandLine(args)
+                .AddEnvironmentVariables(prefix: WebHostDefaults.EnvironmentVariablesPrefix)
+                .AddJsonFile(WebHostDefaults.HostingJsonFile, optional: true)
```
Comment (Tratcher) [inline samples/SampleStartups/StartupFullControl.cs]: Remove the WebHostDefaults.HostingJsonFile constant, it's no longer used in src.

I'm tempted to remove WebHostDefaults.EnvironmentVariablesPrefix as well, as it's only used once in src.

File: test/Microsoft.AspNetCore.Hosting.Tests/WebHostBuilderTests.cs (hunk 2 above)
Comment (JunTaoLuo) [inline test/Microsoft.AspNetCore.Hosting.Tests/WebHostBuilderTests.cs]: They do, though they will only fail if the deployed test tests something that's environment specific and the failure can be hard to trace back to this issue? Also, I don't think anything tests the hierarchy of the fallbacks, but I assume that it's not deemed important.

Comment (Tratcher) [inline test/Microsoft.AspNetCore.Hosting.Tests/WebHostBuilderTests.cs]: They use the environment to select different Startups for different tests, so yes, they will fail if it's wrong.

Comment (JunTaoLuo) [inline test/Microsoft.AspNetCore.Hosting.Tests/WebHostBuilderTests.cs]: Fair.

File: samples/SampleStartups/StartupHelloWorld.cs [hunk 4]
```diff
@@ -28,7 +28,6 @@ public void Configure(IApplicationBuilder app)
         public static void Main(string[] args)
         {
             var host = new WebHostBuilder()
-                .UseDefaultHostingConfiguration(args)
                 .UseStartup<StartupHelloWorld>()
```
Comment (Tratcher) [inline samples/SampleStartups/StartupHelloWorld.cs]: UseServer?

**Comment by Tratcher** [comment]
:shipit:

File: samples/SampleStartups/StartupHelloWorld.cs (hunk 4 above)
Comment (JunTaoLuo) [inline samples/SampleStartups/StartupHelloWorld.cs]: good catch, here and in `StartupExternallyControlled`.

**Comment by muratg** [comment]
:shipit:

**Pull request merged**